### ISSUE TITLE
self.conns.empty() in Pool.close will not work as expected by big chance

### ISSUE THREAD

**Issue by masknu** [FSX/momoko#141]
When there are busy connections at the time you invoking **[Pool.close()](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L429)**, after **[close_alive()](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L438)** being called, those busy connections will be put back to `self.dead` of ConnectionContainer at next io_loop iteration if connections are managed by **[Pool.manage()](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L339)**. Thus, the **[empty()](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L439)** call next to **[close_alive()](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L438)** would be meaningless.

```python
def close(self):
    if self.closed:
        raise PoolError('connection pool is already closed')

    self.conns.close_alive()
    #following line would be meaningless if there are busy connections out there
    self.conns.empty()
    self.closed = True
```

**Comment by haizaar** [comment]
You are right. However it looks like the issue is even deeper - if I call `Pool.close()` in the middle operation, the retry logic will reestablish the connection, so at the end `Pool.close()` does not do its jobs well.

I think the retry logic should not act if `Pool.closed` is true; and yes, `Pool.close()` should not empty connection sets . 

Also I wonder whether public API should `assert not self.closed` before doing anything. 

What do you think?

**Comment by masknu** [comment]
@haizaar , I agree with your concern. I tried a few changes in momoko to make it work right under any closing time, some changes like:

- Put close check in [Connection.execute](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L731) and [Connection.callproc](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L765) at beginning, raise if disconnected.

- Catch exceptions at `execute()` in [Connection.transaction](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L825) [[connection.L876](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L876), [connection.L879](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L879)], then put sys.exc_info() into [`transaction_future`](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L855). In [_rollback()](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L890), apply the same changes.

- In [Pool.getconn](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L285), check Pool.closed flag before doing anything, if it's closed, raise. And never use [Pool.connect](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L258) second time, it will increase connections. or change it a little bit:
```python
def connect(self):
    future = Future()
    assert self.conns.total <= self.size, "total should not greater than size"
    size = self.size - self.conns.total
    
    #add check here
    if 0 == size and self.conns.all_dead:
        
        def on_getconn(fut):
            try:
                conn = fut.result()
            except Exception as error:
                future.set_exc_info(sys.exc_info())
            else:
                future.set_result(self)
                self.putconn(conn)
                
        self.ioloop.add_future(self.getconn(), on_getconn)
        return future
    
    #change pending count here
    pending = [max(size-1, 0)]

    def on_connect(fut):
        if pending[0]:
            pending[0] -= 1
            return
        # all connection attempts are complete
        if self.conns.dead and self.raise_connect_errors:
            ecp = PartiallyConnectedError("%s connection(s) failed to connect" % len(self.conns.dead))
            future.set_exception(ecp)
        else:
            future.set_result(self)
        log.debug("All initial connection requests complete")
        
    #create connection use altered size
    for i in range(size):
        self.ioloop.add_future(self._new_connection(), on_connect)

    return future
```

After the changes we made, I prefer to use [Pool.getconn](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L285) at the beginning instead of [Pool.conncet](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L258), It could make life simple, change [connection.L316](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L316) in [Pool.getconn](https://github.com/FSX/momoko/blob/master/momoko/connection.py#L285) to:
```python
            if self.conns.total:
                self.ioloop.add_future(self._reanimate(), on_reanimate_done)
            else:
                #new db object, and haven't invoked connect(),start with one connection
                self.ioloop.add_future(self._new_connection(), on_reanimate_done)
```

**Comment by haizaar** [comment]
Frankly I'm a bit swamped and hardly will get to work on it until middle next week. If you want to submit PRs with unittests for each of the issues, I'll be happy to review and merge them. Do you want to give it a shot?
